config/config.go

```diff
@@ -1530,115 +1468,103 @@ func migrateConfig(configFile, targetDir string) (string, error) {
 		return nil
 	}
 
-	if c.EncryptConfig == fileEncryptionPrompt {
-		confirm, err := promptForConfigEncryption()
-		if err != nil {
-			log.Errorf(log.ConfigMgr, "The encryption prompt failed, ignoring for now, next time we will prompt again. Error: %s\n", err)
-			return nil
-		}
-		if confirm {
-			c.EncryptConfig = fileEncryptionEnabled
-			return c.SaveConfigToFile(defaultPath)
-		}
+	return c.saveWithEncryptPrompt(path)
+}
 
-		c.EncryptConfig = fileEncryptionDisabled
-		err = c.SaveConfigToFile(defaultPath)
-		if err != nil {
-			log.Errorf(log.ConfigMgr, "Cannot save config. Error: %s\n", err)
+// readConfig loads config from a io.Reader into the config object
+// versions manager will upgrade/downgrade if appropriate
+// If encrypted, prompts for encryption key
+func (c *Config) readConfig(d io.Reader, keyProvider keyProvider) error {
+	j, err := io.ReadAll(d)
+	if err != nil {
+		return err
+	}
+
+	if IsEncrypted(j) {
+		if j, err = c.decryptConfig(j, keyProvider); err != nil {
+			return err
 		}
 	}
+
+	if j, err = versions.Manager.Deploy(context.Background(), j); err != nil {
+		return err
+	}
+
+	if err := json.Unmarshal(j, c); err != nil {
+		return err
+	}
+
 	return nil
 }
 
-// ReadConfig verifies and checks for encryption and loads the config from a JSON object.
-// Prompts for decryption key, if target data is encrypted.
-// Returns the loaded configuration and whether it was encrypted.
-func ReadConfig(configReader io.Reader, keyProvider func() ([]byte, error)) (*Config, bool, error) {
-	reader := bufio.NewReader(configReader)
-
-	pref, err := reader.Peek(len(EncryptConfirmString))
-	if err != nil {
-		return nil, false, err
+// saveWithEncryptPrompt will prompt the user if they want to encrypt their config
+// If they agree, c.EncryptConfig is set to Enabled, the config is encrypted and saved
+// Otherwise, c.EncryptConfig is set to Disabled and the file is resaved
+func (c *Config) saveWithEncryptPrompt(path string) error {
+	if confirm, err := promptForConfigEncryption(); err != nil {
+		return nil //nolint:nilerr // Ignore encryption prompt failures; The user will be prompted again
+	} else if confirm {
+		c.EncryptConfig = fileEncryptionEnabled
+		return c.SaveConfigToFile(path)
 	}
 
-	if !ConfirmECS(pref) {
-		// Read unencrypted configuration
-		decoder := json.NewDecoder(reader)
-		c := &Config{}
-		err = decoder.Decode(c)
-		return c, false, err
+	c.EncryptConfig = fileEncryptionDisabled
+	if err := c.SaveConfigToFile(path); err != nil {
+		return err
 	}
 
-	conf, err := readEncryptedConfWithKey(reader, keyProvider)
-	return conf, true, err
+	return nil
 }
 
-// readEncryptedConf reads encrypted configuration and requests key from provider
-func readEncryptedConfWithKey(reader *bufio.Reader, keyProvider func() ([]byte, error)) (*Config, error) {
-	fileData, err := io.ReadAll(reader)
-	if err != nil {
-		return nil, err
-	}
+type keyProvider func() ([]byte, error)
```

Comment:
I'd like to keep it next to usage, since it's only used in these 2 funcs.
It should be above `readConfig` though!

Fixed a70c38ac5

Discussion:
TL;DR: This type is not the same as the exported/shared types in config_types.go and can live here, and that's idiomatic.
I think I've mentioned before, but IMHO `_types.go` files are mostly idiosyncratic; There's cases where they make sense for a large collection of types related to an exchange that hasn't been broken down, but it's more idiomatic to have organisation where the types and their related functions live in separate files. That's how it is in Go source, unless I've missed any examples, and popular projects ( docker, kubernetes, hugo, gin, prometheus )
The bulk of config types are fine in `_types` because they really are shared types without function, and would overwhelm a standard module package.